crates/rome_formatter/src/lib.rs

```diff
@@ -152,26 +154,254 @@ impl Default for FormatOptions {
     }
 }
 
+/// Lightweight sourcemap marker between source and output tokens
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SourceMarker {
+    /// Position of the marker in the original source
+    pub source: TextSize,
+    /// Position of the marker in the output code
+    pub dest: TextSize,
+}
+
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub struct Formatted {
     code: String,
+    range: Option<TextRange>,
+    sourcemap: Vec<SourceMarker>,
 }
 
 impl Formatted {
-    pub fn new(code: String) -> Self {
-        Self { code }
+    fn new(code: String, range: Option<TextRange>, sourcemap: Vec<SourceMarker>) -> Self {
+        Self {
+            code,
+            range,
+            sourcemap,
+        }
+    }
+
+    fn empty() -> Self {
+        Self {
+            code: String::new(),
+            range: None,
+            sourcemap: Vec::new(),
+        }
+    }
+
+    /// Range of the input source file covered by this formatted code,
+    /// or None if the entire file is covered in this instance
+    pub fn range(&self) -> Option<TextRange> {
+        self.range
+    }
+
+    pub fn sourcemap(&self) -> &[SourceMarker] {
+        &self.sourcemap
     }
 
     pub fn code(&self) -> &String {
         &self.code
     }
+
+    pub fn into_code(self) -> String {
+        self.code
+    }
 }
 
 /// Formats a JavaScript (and its super languages) file based on its features.
 ///
 /// It returns a [Formatted] result, which the user can use to override a file.
 pub fn format(options: FormatOptions, syntax: &SyntaxNode) -> FormatResult<Formatted> {
-    Formatter::new(options).format_root(syntax)
+    let element = Formatter::new(options).format_root(syntax)?;
+    Ok(Printer::new(options).print(&element))
+}
+
+/// Formats a range withing a JavaScript file
```

Comment:
```suggestion
/// Formats a range withing a file, supported by Rome
```

I presume this API will be generic and used to format other files

Reply:
I think we should also mention in which case the API can fail to format some code.

Reply:
As the failure behavior of the formatter might be subject to change I'm not sure if we want to document it just yet. It's something that could go into the crate-level documentation since it's the same for all entry points of the formatter: currently formatting never directly fails, but upon encountering missing or unknown nodes it may print the input tokens verbatim instead of reformatting them